### PR TITLE
Vote Transactions carry the last N votes, which blows up their size.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Vote transactions are large.

#### Summary of Changes

The vote should just carry the last bank hash, and its Tower stack of slots.

Fixes #

tag: @carllin @sagar-solana @rob-solana

### CODE REVIEW

File: programs/vote_api/src/vote_state.rs [hunk 1]
```diff
@@ -148,26 +148,16 @@ impl VoteState {
...
-        if self
-            .votes
-            .back()
-            .map_or(false, |old_vote| old_vote.slot >= vote.slot)
-        {
+    pub fn process_vote(&mut self, votes: &Vote, slot_hashes: &[(Slot, Hash)], epoch: Epoch) {
+        if vote.slots.is_empty() {
             return;
         }
+        let last_slot = -vote.slots.last().unwrap();
```
Comment (sagar-solana) [inline programs/vote_api/src/vote_state.rs]: why the "-" here?

Comment (aeyakovenko) [reply programs/vote_api/src/vote_state.rs]: typo

**Comment by aeyakovenko** [comment]
@carllin if you are going to implement slashing the way we discussed, i think it can use this Vote formatl

**Comment by aeyakovenko** [comment]
@carllin @sagar-solana @rob-solana any idea why `test_ledger_cleanup_service` would fail?

**Comment by sagar-solana** [comment]
There's a ton of dropped votes. Most local cluster tests are disabled otherwise it's likely others would fail as well.

File: programs/vote_api/src/vote_state.rs [hunk 2]
```diff
@@ -147,52 +147,72 @@ impl VoteState {
             }
         }
     }
+    pub fn check_slots_are_valid(&self, vote: &Vote, slot_hashes: &[(Slot, Hash)]) -> bool {
```
Comment (rob-solana) [inline programs/vote_api/src/vote_state.rs]: does this need to be pub ?

File: programs/vote_api/src/vote_state.rs [hunk 3]
```diff
@@ -770,17 +791,25 @@ mod tests {
...
+    #[test]
+    fn test_process_vote_skips_old_vote() {
+        let mut vote_state = VoteState::new(&Pubkey::default(), &Pubkey::default(), 0);
+
+        let vote = Vote::new(vec![0], Hash::default());
+        let slot_hashes: Vec<_> = vec![(*vote.slots.last().unwrap() + 1, vote.hash)];
+        vote_state.process_vote(&vote, &slot_hashes, 0);
+        assert!(recent_votes(&vote_state).is_empty());
+    }
+
```
Comment (rob-solana) [inline programs/vote_api/src/vote_state.rs]: can you write a test that isolates your new function (check_slots_are_valid) ?

**Comment by carllin** [comment]
That failure makes me think that we're probably running into the `No next leader found` issue, so the node will be unresponsive, hence the failure. 

This is probably because for some reason, a root is not being set in the first 96 slots? Not sure if caused by this PR or just a symptom of the  lingering issues in local cluster that we haven't looked into since our last quarterly local cluster cleanup effort :P

**Comment by aeyakovenko** [comment]
@rob-solana there is away to reduce the size further.  The vote that is constructed can check against the bank it's voting on, and only send the missing locktower slots.

File: core/src/consensus.rs [hunk 4]
```diff
@@ -165,24 +164,47 @@ impl Tower {
...
+        hash: Hash,
+        bank_vote_state: &VoteState,
+    ) -> Vote {
+        let mut local_vote_state = local_vote_state.clone();
+        let vote = Vote {
+            slots: vec![slot],
+            hash,
+        };
+        local_vote_state.process_vote_unchecked(&vote);
+        let bank_slots = local_vote_state
```
Comment (sagar-solana) [inline core/src/consensus.rs]: Is `local_vote_state.votes.len().max(bank_vote_state.votes.len())` better than an iter + zip? defer to you judgement.

File: core/src/consensus.rs [hunk 5]
```diff
@@ -195,9 +217,16 @@ impl Tower {
...
+        let vote = Vote {
+            slots: vec![slot],
+            hash,
+        };
+        self.record_bank_vote(vote)
+    }
 
-    pub fn recent_votes(&self) -> Vec<Vote> {
-        self.recent_votes.iter().cloned().collect::<Vec<_>>()
+    pub fn recent_vote(&self) -> Vote {
```
Comment (sagar-solana) [inline core/src/consensus.rs]: we might as well call this `last_vote` now.

**Comment by aeyakovenko** [comment]
@sagar-solana @carllin 

https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L298

I am seeing the above warning, any idea what is causing it?  I don't really follow how these changes would cause it.

**Pull request merged**